federation/pkg/federation-controller/service/dns.go

```diff
@@ -16,25 +16,230 @@ limitations under the License.
+	zoneEndpoints = []string{}
+	regionEndpoints = []string{}
+	globalEndpoints = []string{}
+	zoneName, err := s.getClusterZoneName(clusterName)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	regionName, err := s.getClusterRegionName(clusterName)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	for lbClusterName, lbStatus := range cachedService.serviceStatusMap {
+		lbZoneName, err := s.getClusterZoneName(lbClusterName)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		lbRegionName, err := s.getClusterRegionName(lbClusterName)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		for _, ingress := range lbStatus.Ingress {
+			var address string
+			// We should get either an IP address or a hostname - use whichever one we get
+			if ingress.IP != "" {
+				address = ingress.IP
+			}
+			if ingress.Hostname != "" {
+				address = ingress.Hostname
+			}
+			if len(address) <= 0 {
+				return nil, nil, nil, fmt.Errorf(
+					"Service %s/%s in cluster %s has neither LoadBalancerStatus.ingress.ip nor LoadBalancerStatus.ingress.hostname."+
+						"Cannot use it as endpoint for federated service.",
+					cachedService.lastState.Name, cachedService.lastState.Namespace, clusterName)
+			}
+			if lbZoneName == zoneName {
+				zoneEndpoints = append(zoneEndpoints, address)
+			}
+			if lbRegionName == regionName {
+				regionEndpoints = append(regionEndpoints, address)
+			}
+			globalEndpoints = append(globalEndpoints, address)
+		}
+	}
+	return zoneEndpoints, regionEndpoints, globalEndpoints, nil
+}
+
 // getClusterZoneName returns the name of the zone where the specified cluster exists (e.g. "us-east1-c" on GCE, or "us-east-1b" on AWS)
-func getClusterZoneName(clusterName string) string {
+func (s *ServiceController) getClusterZoneName(clusterName string) (string, error) {
 	// TODO: quinton: Get this from the cluster API object - from the annotation on a node in the cluster - it doesn't contain this yet.
-	return "zone-of-cluster-" + clusterName
+	return "zone-of-cluster-" + clusterName, nil
 }
 
 // getClusterRegionName returns the name of the region where the specified cluster exists (e.g. us-east1 on GCE, or "us-east-1" on AWS)
-func getClusterRegionName(clusterName string) string {
+func (s *ServiceController) getClusterRegionName(clusterName string) (string, error) {
 	// TODO: quinton: Get this from the cluster API object - from the annotation on a node in the cluster - it doesn't contain this yet.
-	return "region-of-cluster-" + clusterName
+	return "region-of-cluster-" + clusterName, nil
 }
 
 // getFederationDNSZoneName returns the name of the managed DNS Zone configured for this federation
-func getFederationDNSZoneName() string {
-	return "mydomain.com" // TODO: quinton: Get this from the federation configuration.
+func (s *ServiceController) getFederationDNSZoneName() (string, error) {
+	return "example.com", nil // TODO: quinton: Get this from the federation configuration.
```

Comment:
Note: I'd prefer to fix this in a followup PR.  It needs to come out of federation config, which complicates e2e testing somewhat, and doesn't really impede other integration testing.